src/lib/storage/column_accessor.hpp

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+
+#include "resolve_type.hpp"
+#include "storage/reference_column.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+/**
+ * This is the base class for all ColumnAccessor types.
+ * It provides the common interface to access individual values of a column.
+ */
+template <typename T>
+class BaseColumnAccessor {
+ public:
+  virtual const std::optional<T> access(ChunkOffset offset) const = 0;
+  virtual ~BaseColumnAccessor() {}
+};
+
+/**
+ * A ColumnAccessor is templated per ColumnType and DataType (T).
+ * It requires that the underlying column implements an implicit interface:
+ *
+ *   const std::optional<T> get_typed_value(const ChunkOffset chunk_offset) const;
+ *
+ */
+template <typename T, typename ColumnType>
+class ColumnAccessor : public BaseColumnAccessor<T> {
+ public:
+  explicit ColumnAccessor(const ColumnType& column) : _column{column} {}
+
+  const std::optional<T> access(ChunkOffset offset) const final { return _column.get_typed_value(offset); }
+
+ protected:
+  const ColumnType& _column;
```

Comment:
Any reason why this should hold a smart pointer?

Reply:
What do you mean? It does not hold a smart pointer.

If you mean why it does *not* hold a smart pointer, two reasons:
1. I don't see a reason why it should, as it would add indirection without any added use
2. `resolve_column_type` gives a const-ref to the concrete column type and `create_column_accessor` (which is a thin wrapper around that) is the usual way to construct a ColumnAccessor.

Reply:
I meant "not hold", of course. 

We use smart pointers over references for class members everywhere else, so I would have done it here as well for consistency reasons. since `resolve_column_type` doesn't give you a smart pointer to the column, I guess I have to let this one pass